Best Practises UDFs (#374)
fusedio/fused-docs: docs/user-guide/best-practices/udf-best-practices.mdx

---
id: udf-best-practices
title: Making the most out of UDFs
sidebar_label: Building & Running UDFs
sidebar_position: 0
---

# Build & Running UDFs

_An opinionated guide to making the most out of Fused UDFs_

[Fused UDFs](/core-concepts/write/) are Python functions that run on serverless compute and can be called from anywhere with [`fused.run(udf)`](/core-concepts/run-udfs/run-small-udfs/#fusedrun). This guide is a resource meant to help you on your way to making the most out of UDFs.

## A short reminder: The anatomy of a UDF

```python showLineNumbers
@fused.udf
def udf(my_awesome_input: int = 1):
    import pandas as pd

    return pd.DataFrame({"Look at my output: ": [my_awesome_input]})
```

Each UDF has a few specific elements:
- The [`@fused.udf` decorator](/core-concepts/write/#fusedudf-decorator)
- Arguments -ideally typed-

> - imports _inside_ the function

Reviewer: ```suggestion
- Imports _inside_ the function
```

(just for being consistent)

- Some logic
- A [supported `return` object](/core-concepts/write/#return-object)

:::note

> All of this is explained in the ["Write UDF"](/core-concepts/write/) section in much more details

Reviewer: ```suggestion
All of this is explained in the ["Write UDF"](/core-concepts/write/) section in much more details.
```

:::

> You can then run UDFs from _anywhere_ you have access to Python with [`fused.run(udf)`](/core-concepts/run-udfs/run-small-udfs/#fusedrun). These are still Python functions, giving you a lot of flexibility on what oyu can do, but we have some recommendations for keeping them fast & efficient.

Reviewer: ```suggestion
You can then run UDFs from _anywhere_ where you have access to Python with [`fused.run(udf)`](/core-concepts/run-udfs/run-small-udfs/#fusedrun). These are still Python functions, giving you a lot of flexibility on what oyu can do, but we have some recommendations for keeping them fast & efficient.
```

I am not a native-enough speaker to know what is correct :), but just "from anywhere you have access" reads a bit odd to me ("run from anywhere" is a complete sentence, so I don't expect it to go on?)

Author: Good point, simplifying

> You can then run UDFs from _anywhere_ you have access to Python with [`fused.run(udf)`](/core-concepts/run-udfs/run-small-udfs/#fusedrun). These are still Python functions, giving you a lot of flexibility on what oyu can do, but we have some recommendations for keeping them fast & efficient.

Reviewer: ```suggestion
You can then run UDFs from _anywhere_ you have access to Python with [`fused.run(udf)`](/core-concepts/run-udfs/run-small-udfs/#fusedrun). These are still Python functions, giving you a lot of flexibility on what you can do, but we have some recommendations for keeping them fast & efficient.
```

## Writing efficient UDFs

### Keep things small

The main benefit of Fused UDFs is how responsive they are. To do so, they run on Python serverless compute, but [quickly timeout](/core-concepts/run-udfs/run-small-udfs/#defining-small-job). The best way to do that is to keep **things** fast is to keep them small:

- Break pipelines into single tasks UDFs
- Leverage [`fused.run()`](/core-concepts/run-udfs/run-small-udfs/#fusedrun) to chain together UDFs

<details>
    <summary>Example: Breaking down a complex pipeline into smaller UDFs</summary>

    ❌ Not recommended:

    ```python showLineNumbers
    @fused.udf
    def inefficient_pipeline_udf(data_path):
        import pandas as pd

        df = pd.read_csv(data_path)
        # Some complicated processing logic to create df_processed
        processed_df = ... 

>         return processed_df

Reviewer: I certainly understand what we try to show here, I think we just have to be careful about the fact that breaking it up into multiple steps just like this, also adds overhead because we now have to pass the dataframe from one step to another. If you run such pipeline locally that is of course not an issue and breaking it up this way is perfect. But with our remote compute, the dataframe actually needs be serialized -> send -> deserialize and then pass to second udf and again serialize -> send -> deserialize when running the second UDF. 
So doing something like the example here with a somewhat larger `df` that gets passed from one to the other function might actually be slower

In practice, with a larger dataframe, you probably shouldn't return it as a result, but let the first UDF write it to /mount as Parquet, and then let the second step read that again. Or write the CSV reading step as a cached sub-function. 
But of course that goes into more detailed concepts than what we want to show here...

Author: That's a valid point

I'm not going to change this for now, but make a note of this

    ```

    ✅ Instead, break it down:

    ```python showLineNumbers
    @fused.udf
    def load_data_udf(data_path):
        import pandas as pd
        return pd.read_csv(data_path)
    ```

    ```python showLineNumbers
    @fused.udf
    def process_data_udf(df):
        import pandas as pd

        # Some complicated processing logic to create df_processed
        processed_df = ... 

        return processed_df
    ```

    ```python showLineNumbers
    @fused.udf
    def pipeline_udf(data_path):
        import pandas as pd

        df = fused.run(load_data_udf, data_path=data_path)
        processed_df = fused.run(process_data_udf, df=df)

        return processed_df
    ```
</details>

### Run often, Iterate quickly

Just like writing short cells when developing in a Jupyter Notebook, we recommend you keep your UDFs short & fast to execute

> ⚡️ Aim for **UDFs that take 30s - 1min to run**

Reviewer: (for a map context with tiles, you probably even want to aim at a few seconds .. but this of course depends a lot on the context)

Author: Good point

Though caching does change this a bit, as run number 1 might be longer than run number 2
But you're right, changing this to "up to 1min" instead

Reviewer: > Though caching does change this a bit, as run number 1 might be longer than run number 2
Good point, you can have a snappy map experience if all the tiles you are looking at were already viewed recently

> UDFs run with [`fused.run()`](/core-concepts/run-udfs/run-small-udfs/#fusedrun) [timeout after 120s](/core-concepts/run-udfs/run-small-udfs/#defining-small-job) so we recommend you keep a buffer in case your UDF takes a bit longer to execute

Reviewer: ```suggestion
UDFs run with [`fused.run()`](/core-concepts/run-udfs/run-small-udfs/#fusedrun) [time out after 120s](/core-concepts/run-udfs/run-small-udfs/#defining-small-job) so we recommend you keep a buffer in case your UDF takes a bit longer to execute
```

<details>
    <summary>Visual: UDF timing guideline</summary>

    This is a breakdown of what happens when you run a UDF with `fused.run()` and why we recommend you keep your UDFs at the 30s-1min mark:

    ![UDF Design Guidelines](/img/user-guide/best-practices/udf_design_timing.png)

</details>

### But what if I want a longer UDF?

A lot of processing of large datasets sometimes doesn't fit in a 30s-1min job and needs to run for longer. You have a few options:
- Can you break your pipeline into [smaller UDFs?](/user-guide/best-practices/udf-best-practices/#keep-things-small)
- You can run [multiple small UDF in parallel](/core-concepts/run-udfs/run-small-udfs/#running-multiple-small-fusedrun-jobs-in-parallel)

> - If you need a much longer or simply need more RAM [run you can a large UDF with our offline instances](/core-concepts/run-udfs/run_large/)

Reviewer: ```suggestion
- If you need a much longer run or simply need more RAM, [you can run a large UDF with our offline instances](/core-concepts/run-udfs/run_large/)
```

### [Cache](/core-concepts/cache/) as much as you can

Fused relies heavily on [caching](/core-concepts/cache/) repetitive tasks to make recurring calls much faster (and more compute efficient)

✅ You want to use caching for functions with inputs that are recurring:

> - Loading a dataset

Reviewer: In practice you only want that if it is loading a dataset from a less efficient format like CSV? I would guess that reading a Parquet file directly vs getting the cached result might not make much of a difference?

Reviewer: Ah, that probably depends on the source. If reading it from slow s3 vs fast local cache, of course then also caching the parquet read is useful (but for reading parquet from /mount it might not be useful)

Author: Yeah I'm keeping it rule of thumb here rather than absolutes. So things that are true in 90% of the cases, especially for people who aren't on top of the latest file formats

Parquet might still be slow-ish if it's not partitioned for example (I think?)

Reviewer: > Parquet might still be slow-ish if it's not partitioned for example (I think?)

Yes, good point, especially for a tile UDF, reading from a non-spatially partitioned parquet file can be very expensive and a good use case for caching (although we of course would also say people should then use our ingest method to make it faster). 

I assume that even if you have a spatially partitioned parquet dataset, depending on the tile you are looking at it might not align exactly with one of the chunks of the parquet dataset, and so you might be reading too much and then filtering it down to just the tile. Also for those cases caching might still make sense. 

Of course, if the full UDF gets cached, it is not necessarily needed to cache the reading subset of the UDF (at least when the UDF is stable, while you are developing it and the UDF is still changing, then caching the reading part separately could be very useful)

- Computing a recurring operation with default variables 
- Intermediate results you'll reuse soon

❌ When not to use caching:

> - In most cases, for functions taking `bbox` as an argument -> your function + input would get overwritten for each new `bbox` which probably changes for each new call

Reviewer: The "get overwritten" is strange in this context (what gets overwritten? the cache doesn't get "overwritten" when calling it with a new param, it's just not using the cache then / creating new cache entries each time and not fetching from the cache)

Author: Rephrased this

<details>
    <summary>Example: Caching a repetitive task</summary>

    Re-using the example from [keeping things small](/user-guide/best-practices/udf-best-practices/#keep-things-small):

    ❌ Not recommended:

    ```python showLineNumbers
    @fused.udf
    def inefficient_pipeline_udf(data_path):
        import pandas as pd

        df = pd.read_csv(data_path)
        # Some complicated processing logic to create df_processed
        processed_df = ... 

        return processed_df
    ```

    ✅ Instead, break it down AND cache the calls:

    ```python showLineNumbers
    @fused.udf
    def load_data_udf(data_path):
        import pandas as pd
        return pd.read_csv(data_path)
    ```

    ```python showLineNumbers
    @fused.udf
    def process_data_udf(df):
        import pandas as pd
        # Some complicated processing logic to create df_processed
        # ...
        return processed_df
    ```

    ```python {5-7,9-11} showLineNumbers
    @fused.udf
    def pipeline_udf(data_path):
        import pandas as pd

        @fused.cache
        def load_data(data_path):

>             return fused.run(load_data_udf, data_path=data_path)

Reviewer: You could also put the `pd.read_csv(data_path)` here directly instead of wrapping it in a udf and calling `fused.run` here (essentially like the linked example in https://docs-staging.fused.io/pr/374/core-concepts/data_ingestion/why-ingestion/#using-cache-as-a-single-use-ingester does). Do we have any ideas about which pattern is best or preferred?

Author: As you mentioned, `fused.run()` does have a hit on moving `df` around

But I'm gonna keep it like this for now to keep things simple

        @fused.cache
        def process_data(df):
            return fused.run(process_data_udf, df=df)

        df = load_data(data_path)
        processed_df = process_data(df)

        return processed_df
    ```
</details>

:::tip
Read more about the caching details:
- [in the dedicated section](/core-concepts/cache/)
- How you can use cache to [speed up exploration of slow to read datasets](/core-concepts/data_ingestion/why-ingestion/#using-cache-as-a-single-use-ingester)
:::

### Prepare your large datasets

> Fused works at its best with data that's fast to read and can be read in tiles or chunks. We know that most of the data out there isn't in the most [efficient file formats](/core-concepts/data_ingestion/file-formats/) which is why we provide tools to [ingest your own data](/core-concepts/data_ingestion/ingestion-your-data/) into cloud-optimized, partitioned formats.

Reviewer: ```suggestion
Fused works at its best with data that is fast to read and can be read in tiles or chunks. We know that most of the data out there isn't in the most [efficient file formats](/core-concepts/data_ingestion/file-formats/) which is why we provide tools to [ingest your own data](/core-concepts/data_ingestion/ingestion-your-data/) into cloud-optimized, partitioned formats.
```

We have a [dedicated page](/core-concepts/data_ingestion/why-ingestion/#when-is-ingestion-needed) for when you should consider ingesting your own data. As a rule of thumb you want to consider ingesting your data when:
- Files are read multiple times and >100MB
- Files that are slow or require some processing to open (`.zip` for example)

### Don't start from scratch: [UDF Catalog](/workbench/udf-catalog/)

> Just like using libraries in Python to leverage existing tools, you don't need to start from scratch in Fused. We have a [Catalog of existing UDFs](/workbench/udf-catalog/) built & maintained by us and the community

Reviewer: ```suggestion
Just like using libraries in Python to leverage existing tools, you don't need to start from scratch in Fused. We have a [Catalog of existing UDFs](/workbench/udf-catalog/) built & maintained by us and the community.
```

You can find a host of different UDFs that can serve as a starting point or as inspiration to create your own UDFs:
- Open datasets from common open repository [like STAC catalogs](https://www.fused.io/workbench/catalog/Sentinel_Tile_Example_2-ddb4c495-40e3-48ba-849c-256487f8a9cb)
- Run [on the fly ML prediction](https://www.fused.io/workbench/catalog/Dl4eo_Airplane_Detection_Global-91f81c0c-08f6-4a0c-82f9-bd2e0322b4e7) on satellite images
- Compute a [spatial join](https://www.fused.io/workbench/catalog/Overture_Nsi-dd89972c-ce30-4544-ba0f-81fc09f5bbef) between 2 datasets

![UDF Catalog](/img/udfcatalog_august.png)

> You can also contribute your own UDFs to the community!

Reviewer: Do we have a link to point to with more details about how to do this?

Author: Not yet, it's a [ticket of mine](https://www.notion.so/fusedio/docs-GitHub-integration-functionality-158899d3b76380aeb6cbc69f36f231ef?pvs=4), that I'm bumping in priority because I think is really important for teams

```python showLineNumbers
@fused.udf
def udf(n: int = 1):
    print(f"{n=}")
    return
```

Since Python 3.8 you can use [f-string debugging](https://docs.python.org/3/whatsnew/3.8.html#f-strings-support-for-self-documenting-expressions-and-debugging) which is what we recommend you use:
```python showLineNumbers
print(f"{my_fancy_variable=}")
```

This allows you to print many variables without getting lost with what is what

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

<Tabs className="unique-tabs">
  <TabItem value="workbench" label="Workbench" default>
    ![UDF Catalog](/img/user-guide/best-practices/workbench_print_stdout.png)
  </TabItem>
  <TabItem value="notebook" label="notebook" default>
    ![UDF Catalog](/img/user-guide/best-practices/notebook_print_stdout.png)
  </TabItem>
</Tabs>

### Type all your inputs

We strongly recommend you [type](https://docs.python.org/3/library/typing.html) all your inputs with the appropriate type:

```python {3} showLineNumbers
@fused.udf
def udf(
    bbox:fused.types.TileGDF=None, n:int=1
):
    ...
    return
```

This has 2 effects:
- It makes your code more readable to others

> - Fused only supports a [few types](/core-concepts/write/#supported-types) at the moment. Any non-typed or unsupported types will be handled as `str`

Reviewer: ```suggestion
- Fused only supports a [few types](/core-concepts/write/#supported-types) at the moment. Any non-typed or unsupported types will be passed as `str`
```

? to make it more clear that the UDF function will receive it as strings (when running remotely)

### Use `time.time()`

Sometimes you're not sure what's taking so long. The simplest way to figure this out is to use [`time.time()`](https://docs.python.org/3/library/time.html#time.time):

<details>
    <summary>Example: finding a slow process</summary>

    ```python {4,9-11,17-19} showLineNumbers
    @fused.udf
    def udf():
        import time
        beginning_time = time.time()
        
        # long processing step #1
        time.sleep(5)
        end_process_1 = time.time()

>         process_time_1 = round(
>             end_process_1 - beginning_time, 2
>         )
>         print(f"{process_time_1=}")

Reviewer: You could also do this:
```suggestion
        process_time_1 = end_process_1 - beginning_time
        print(f"{process_time_1=:.2f}")
```
which is shorter and gives the same result (`.2f` means "format as a float with 2 decimals after the ."). I like this because I am familiar with it ;) But probably for people that don't know it the `round` is much easier to understand (I don't know how widespread this string formatting language is known)

Author: Going for more explicit > shorter here

I also mess up that formatting, so I always find myself reverting to `round()`
